synapse/util/ratelimitutils.py

```diff
@@ -162,6 +166,7 @@ def on_wait_finished(_: Any) -> "defer.Deferred[None]":
 
         def on_start(r: object) -> object:
             logger.debug("Ratelimit [%s]: Processing req", id(request_id))
+            wait_span_scope.__exit__(None, None, None)
```

Comment:
I'm a bit concerned that this approach could lead to us forgetting to exit the scope in some situations (particularly: on cancellation, or other exceptions such as the place where we `raise LimitExceededError`).

one approach we've taken elsewhere is to make the span cover the whole of the limited operation (ie, the wait, *and* the operation itself), and just emit a tracing event when the limiter completes (or have a nested span). That means you can just do a `with start_active_span(...)` as normal.

Failing that, I'd suggest doing this in `on_both` instead of `on_start`, and moving the call to `start_active_span` to just before the definition of `on_both`, to minimise the amount of code that can throw exceptions between the two.

Reply:
> one approach we've taken elsewhere is to make the span cover the whole of the limited operation (ie, the wait, and the operation itself), and just emit a tracing event when the limiter completes (or have a nested span). That means you can just do a with `start_active_span(...)` as normal.

We already have a span that covers the wait and the operation itself. I'm interested in exposing how long the wait is though.

---

> Failing that, I'd suggest doing this in `on_both` instead of `on_start`, and moving the call to `start_active_span` to just before the definition of `on_both`, to minimise the amount of code that can throw exceptions between the two.

Moved to this 👍

Reply:
> I'm interested in exposing how long the wait is though.

doesn't emitting a tracing event do that though?

> Moved to this

ok then :)